vispy/scene/widgets/widget.py

```diff
@@ -182,21 +186,37 @@ def _update_line(self):
         #  ........
         #  ........
         #
+        l = b = m
         r = int(self.size[0]) - m
         t = int(self.size[1]) - m
-        
         pos = np.array([
-            [m, m], [m+1, m+1],
-            [r, m], [r-1, m+1],
-            [r, t], [r-1, t-1],
-            [m, t], [m+1, t-1],
-            [m, m], [m+1, m+1]
+            [l, b], [l+w, b+w],
+            [r, b], [r-w, b+w],
+            [r, t], [r-w, t-w],
+            [l, t], [l+w, t-w],
         ], dtype=np.float32)
-        
-        self._visual.set_data(vertices=pos)
+        faces = np.array([
+            [0, 2, 1],
+            [1, 2, 3],
+            [2, 4, 3],
+            [3, 5, 4],
+            [4, 5, 6],
+            [5, 7, 6],
+            [6, 0, 7],
+            [7, 0, 1],
+            [5, 3, 1],
```

Comment:
Last two triangles should be left out if the background color is clear. I have seen big performance hits for filling large triangles like these.